incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/config/GrpcKafkaWithResolver.java

```diff
@@ -156,12 +154,33 @@ public GrpcKafkaWithProduceResult resolveProduce(
         String16FW topic = new String16FW(produce.topic);
         KafkaAckMode acks = produce.acks;
 
-        final GrpcMetadataFW idempotencyKey = beginEx.metadata().matchFirst(m ->
-            GRPC_METADATA_NAME_IDEMPOTENCY_KEY.value().compareTo(m.name().value()) == 0);
+        Array32FW<GrpcMetadataFW> metadata = beginEx.metadata();
 
         final String16FW service = new String16FW(beginEx.service().asString());
         final String16FW method = new String16FW(beginEx.method().asString());
 
+        OctetsFW correlationId = resolveCorrelationId(metadata);
+        Supplier<DirectBuffer> keyRef = resolveKey(authorization, produce);
+
+        GrpcKafkaWithProduceHash hash = new GrpcKafkaWithProduceHash(octetsRW, dashOctetsRW, correlationId, hashBytesRW);
+        hash.digestHash();
+        hash.updateHash(beginEx.service().value());
+        hash.updateHash(beginEx.method().value());
+        hash.updateHash(metadata.items());
```

Comment:
Why do we pre-hash these here?

Shouldn't we just hash all the headers of the kafka message when we're producing, which includes these values already?